rslib/src/scheduler/fsrs/weights.rs

```diff
@@ -95,40 +105,63 @@ fn anki_to_fsrs(revlogs: Vec<RevlogEntry>, next_day_at: TimestampSecs) -> Vec<FS
     revlogs
 }
 
-fn single_card_revlog_to_items(
-    mut entries: Vec<RevlogEntry>,
+/// When updating memory state, FSRS only requires the last FSRSItem that
+/// contains the full history.
+pub(super) fn fsrs_items_for_memory_state(
+    revlogs: Vec<RevlogEntry>,
     next_day_at: TimestampSecs,
-) -> Option<Vec<FSRSItem>> {
-    // Find the index of the first learn entry in the last continuous group
-    let mut index_to_keep = 0;
-    let mut i = entries.len();
-
-    while i > 0 {
-        i -= 1;
-        if entries[i].review_kind == RevlogReviewKind::Learning {
-            index_to_keep = i;
-        } else if index_to_keep != 0 {
-            // Found a continuous group
-            break;
+) -> Vec<(CardId, FSRSItem)> {
+    let mut out = vec![];
+    for (card_id, group) in revlogs.into_iter().group_by(|r| r.cid).into_iter() {
+        let entries = group.into_iter().collect_vec();
+        if let Some(mut items) = single_card_revlog_to_items(entries, next_day_at) {
+            if let Some(item) = items.pop() {
+                out.push((card_id, item));
+            }
         }
     }
+    out
+}
 
-    // Remove all entries before this one
-    entries.drain(..index_to_keep);
-
-    // we ignore cards that don't start in the learning state
-    if let Some(entry) = entries.first() {
-        if entry.review_kind != RevlogReviewKind::Learning {
-            return None;
+/// Transform the revlog history for a card into a list of FSRSItems. FSRS
+/// expects multiple items for a given card when training - for revlog
+/// `[1,2,3]`, we create FSRSItems corresponding to `[1,2]` and `[1,2,3]`.
+fn single_card_revlog_to_items(
```

Comment:
We can pass in an arg to control whether to send out [[1], [1,2], [1,2,3]] or [[1,2], [1,2,3]].

Reply:
For calculating, we need to send out [1] for revlog [1].